File: src/geek time/c#/week03/236. Lowest Common Ancestor of a Binary Tree.cs
using System;

namespace week03 {
    public partial class Solution {
        public TreeNode LowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
            // 自己做的时候一直想不到怎么同时拥有孩子和父亲节点的判定场景
            // 解法代码中给出了答案，当左右都存在时，说明第一个祖先就是自己
            if (root == null || root == p || root == q) {
                return root;
            }
            TreeNode left = LowestCommonAncestor(root.left, p, q);
            TreeNode right = LowestCommonAncestor(root.right, p ,q);
            return left != null && right != null ? root : left == null ? right : left;
        }
    }
}

File: src/geek time/typescript/week03/236. Lowest Common Ancestor of a Binary Tree.ts
export class TreeNode {
    public val: number;
    public left: TreeNode;
    public right: TreeNode;
    constructor(val: number) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}
function lowestCommonAncestor(root: TreeNode, p: TreeNode, q: TreeNode): TreeNode {
    if (!root || root === p || root === q) return root;
    const left = lowestCommonAncestor(root.left, p, q);
    const right = lowestCommonAncestor(root.right, p, q);
    return left !== null ? (right !== null ? root : left) : right;
}
